feat: pluggable test runner

imrekoszo opened the pull request:
Add the option to supply a custom test runner.

A previous version of this PR contained an example implementation using kaocha, I will recreate that outside of this repo and link it.

Todo:

- documentation
  - [x] document plugin and constructor interface
  - [x] document how to pass extra args to test runner using user-input (focus) and stuff like verbose
- [x] create repo for kaocha impl that was pulled from this PR
  - https://github.com/imrekoszo/polylith-kaocha
- [x] create issue about namespaces vs paths
- later perhaps
  - [ ] extend info integration for test discovery? potentially with a flag that hides a potentially slow `tests-present?` 
    - this is proving to be somewhat difficult as the source data for this gets collected rather early: https://github.com/polyfy/polylith/blob/e874e809aa9d3cecaa0540376bd2875f99b37372/components/workspace/src/polylith/clj/core/workspace/project.clj#L40-L50

kkrausse commented on components/kaocha-test-runner-example/src/polylith/clj/core/kaocha_test_runner_example/core.clj:
```diff
@@ -0,0 +1,88 @@
+       (and (test-runner-plugin/test-sources-present? this)
+            (let [config (->test-plan-config eval-in-project)]
+              (eval-in-project
+               `(do (require ~''kaocha.api ~''kaocha.testable ~''kaocha.hierarchy)
```
I don't think the `~'` is required here. And same with below. I don't think syntax quote auto-namespaces symbols with dots (or slashes) in them

imrekoszo replied:
You're right, I'm removing them.

imrekoszo committed:
Add pluggable test runner

imrekoszo committed:
Add note about class loader error reporting

imrekoszo committed:
Example: implementation of test runner plugin

imrekoszo committed:
Example: use kaocha test runner for poly project

imrekoszo committed:
Save discussion comments

imrekoszo committed:
Fix set member check

imrekoszo committed:
Break up test-runner namespace

- extract pluggable test runner impl
- extract some common message ops
- leave only the responsibility of organizing the high level process

imrekoszo committed:
Fix misplaced docstring

imrekoszo committed:
Stop being needlessly lazy in a few places

imrekoszo committed:
Remove unused argument

imrekoszo committed:
Fix missing test path

imrekoszo committed:
Support global test runner config

imrekoszo committed:
More lazy->transducer refactor

imrekoszo committed:
Try out the new workspace shema and test runner config a bit more

imrekoszo committed:
Extract test runner init

imrekoszo committed:
Integrate with poly check, add tests

imrekoszo committed:
Remove references to kaocha

imrekoszo committed:
Cleanup formatting

imrekoszo committed:
Cleanup todos, improve docstring

imrekoszo committed:
Fix workspace tests

imrekoszo committed:
Add project-settings tests

imrekoszo committed:
Integrate with help check

imrekoszo committed:
Merge test-runner-plugin with -init, remove trp project

imrekoszo committed:
Make default-test-runner a component, fix tests

imrekoszo committed:
Some more lazy->transducer refactoring

imrekoszo committed:
Rename components, keys and namespaces as agreed

imrekoszo committed:
Throw when unable to resolve symbol to var

imrekoszo committed:
Default test runner to be assigned at config time

furkan3ayraktar commented on components/clojure-test-test-runner/src/polylith/clj/core/clojure_test_test_runner/core.clj:
```diff
@@ -0,0 +1,97 @@
+        {:keys [name namespaces paths]} project
+        {:keys [project-to-bricks-to-test project-to-projects-to-test]} changes
+
+        ;; TODO: if the project tests aren't to be run, we might further narrow this down
```
Is this something we will address now or a TODO for after we merge this PR?

imrekoszo replied:
I intended this as a note for later as I don't want to change what the default test runner does as part of this effort.

furkan3ayraktar replied:
Alright, then we can leave this TODO for later 👍

furkan3ayraktar commented on components/common/src/polylith/clj/core/common/class_loader.clj:
```diff
@@ -72,5 +72,6 @@
       (eval-in* class-loader '(require 'clojure.main))
       class-loader)
     (catch Exception e
+      ;; TODO: this seems to not print an error when clojure.lang.RT couldn't load
```
Hmm, two questions here:
1. In which case `clojure.lang.RT` could not load?
2. Could we print an error in a different way here so that we do not loose the error message? If we can't maybe we can just turn this into a comment rather than a TODO

imrekoszo replied:
I found this while iterating on an early version of the kaocha runner implementation and had an error in the kaocha-specific code which manifested as the load error. What I noticed that the error wasn't getting printed.

Let me try and set up a repro case somewhere.

imrekoszo replied:
Haven't managed to repro this, it was probably not a good idea to not commit a failing setup when I came across it.

In any case, the error that happened here was caused by me setting up something wrong. For now I removed this comment and will raise an issue with a repro case if I come across it again.

furkan3ayraktar replied:
Sounds good

furkan3ayraktar commented on components/workspace-clj/src/polylith/clj/core/workspace_clj/project_settings.clj:
```diff
@@ -1,12 +1,24 @@
+
+(defn or-default-constructor [candidate]
+  (if (contains? #{nil :default} candidate)
+    'polylith.clj.core.clojure-test-test-runner.interface/create
```
Since this is a symbol, in case of a refactoring we could easily forget this. Also Polylith will not understand that `workspace-clj` actually depends on `clojure-test-test-runner`.

How could we overcome this? Could we maybe just require `clojure-test-test-runner` in this namespace?

imrekoszo replied:
We could require for sure, but that would mean it will load that namespace even when someone does not want to use `clojure-test-test-runner`. As long as the polylith workspace uses the default test runner assigned here, `poly check` and `poly test` will both fail if this gets outdated.

furkan3ayraktar replied:
Alright, based on our discussion on Slack, this should be okay to be left as it is and an explicit test could be added to the `poly` project.

imrekoszo replied:
Added in https://github.com/polyfy/polylith/pull/196/commits/68af10c8e144901e67dc861bd571db6cc32306f1

imrekoszo committed:
Merge remote-tracking branch 'upstream/master' into kaocha

imrekoszo committed:
Remove todo comment

Was unable to reproduce the issue

imrekoszo committed:
Add test runner info to `help test`

imrekoszo committed:
Make help test text more precise.

imrekoszo committed:
Bump patch version & revision & date

imrekoszo committed:
Test to ensure default test runner ships with poly

imrekoszo committed:
Add note about user input to test runner docstring

imrekoszo committed:
Bump revision

imrekoszo committed:
Support multiple test runners per project

imrekoszo committed:
Bump revision

imrekoszo committed:
Add multiple test runners info to `help test`, bump revision

imrekoszo committed:
Make test runner errors more visible

imrekoszo committed:
Merge remote-tracking branch 'upstream/master' into kaocha

tengstrand commented on components/workspace/src/polylith/clj/core/workspace/core.clj:
```diff
@@ -31,14 +30,14 @@
         suffixed-top-ns (common/suffix-ns-with-dot top-namespace)
         interfaces (interfaces/calculate components)
-        interface-names (apply sorted-set (filter identity (mapv :name interfaces)))
+        interface-names (into (sorted-set) (keep :name) interfaces)
```
Awesome that you find and improve things like this!
